integration_tests/tests/mod.rs

```diff
@@ -0,0 +1,408 @@
+use structopt::StructOpt;
+use thiserror::Error;
+use zeekoe::customer::database::ClosingBalances;
+
+/// Default balance for establishing channels
+const DEFAULT_BALANCE: u64 = 5;
+/// Default payment; should be less than balance
+const DEFAULT_PAYMENT: u64 = 1;
+
+/// Struct to represent one test: includes a name and a list of operations and outcomes
+#[derive(Debug, Clone)]
+pub struct Test {
+    pub name: String,
+    pub operations: Vec<(Operation, Outcome)>,
+}
+
+/// Set of operations that can be executed by the test harness
+#[allow(unused)]
+#[derive(Debug, Clone, Copy)]
+pub enum Operation {
+    Establish(u64),
+    Pay(u64),
+    PayAll,
+    MutualClose,
+    CustomerClose,
+    MerchantExpiry,
+    Store,
+    Restore,
+    NoOp,
+}
+
+impl Operation {
+    /// Amount of time to wait before validating that an `Operation` has successfully completed.
+    fn wait_time(&self) -> Duration {
+        // The following actions cause delays:
+        // (a) the watcher notices a contract change (60 seconds)
+        // (b) a watcher posts a transaction on chain (10 seconds)
+        // (c) a watcher waits for self-delay to elapse (60 seconds + noticing)
+        let seconds = match self {
+            Self::Establish(_)
+            | Self::Pay(_)
+            | Self::PayAll
+            | Self::Store
+            | Self::Restore
+            | Self::NoOp => 0,
+
+            // The merchant watcher must notice the contract status change
+            Self::MutualClose => 60,
+            // After the initial close tx is posted inline, either:
+            // - the merchant notices, posts Dispute, and waits for it to confirm. The customer
+            //   watcher notices and updates status (130)
+            // - the merchant notices and does nothing. The customer watcher waits for self-delay to
+            //   elapse, then posts claim (130)
+            Self::CustomerClose => 130,
+            // Expiry is either:
+            // - the same as customer close, but preceded by the customer noticing expiry
+            // and posting the corrected balances (+70)
+            // - the merchant waits for self-delay to elapse, then claims funds (130)
+            Self::MerchantExpiry => 200,
+        };
+        Duration::from_secs(seconds)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Outcome {
+    /// Which process, if any, had an error? Assumes that exactly one party will error.
+    error: Option<Party>,
+    /// Outcome of channel; left an Option in case we do not expect a channel from a test
+    /// (e.g. interacting with a non-existent channel).
+    channel_outcome: Option<ChannelOutcome>,
+}
+
+#[derive(Debug, Clone)]
+struct ChannelOutcome {
+    customer_status: CustomerStatus,
+    merchant_status: MerchantStatus,
+    customer_balance: CustomerBalance,
+    merchant_balance: MerchantBalance,
+    /// Closing balances of channels, for use in close-related tests
+    closing_balances: Option<ClosingBalances>,
```

Comment:
thought (non-actionable): It would be nice if this could be a parameter of the close operations, rather than an optional field here. I can't think of a way to do that without significantly changing the structure of `execute()`, though.

Reply:
I think this might make the match a little tricky to get the parameter value out of it. I haven't included this yet, but I'd like to about what a clean way to do this refactor looks like!

Reply:
Yeah, I think it doesn't make sense right now, but we can think about it in the future. There's probably a runtime-error way to enforce it in `execute` where we manually check the enum variant and this field and throw an error if it's not either (a close op, Some) or (a non-close op, None). But it would be handy to find a way to enforce it at compile time.

Anyway, agree that this is a refactor for later.